test/e2e/federated-service.go

```diff
@@ -148,49 +144,67 @@ var _ = framework.KubeDescribe("Service [Feature:Federation]", func() {
-		}
-		for _, name := range svcDNSNames {
-			discoverService(f, name, true)
-		}
-	})
-
-	It("should be able to discover a non-local federated service", func() {
-		framework.SkipUnlessFederated(f.Client)
-
-		createService(f.FederationClientset_1_3, clusterClientSets, f.Namespace.Name)
+	Describe("DNS", func() {
+		BeforeEach(func() {
+			framework.SkipUnlessFederated(f.Client)
+			createService(f.FederationClientset_1_3, clusterClientSets, f.Namespace.Name)
+		})
 
-		// Delete a federated service shard in the default e2e Kubernetes cluster.
-		err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
-		Expect(err).NotTo(HaveOccurred())
-		waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+		It("should be able to discover a federated service", func() {
+			framework.SkipUnlessFederated(f.Client)
 
-		localSvcDNSNames := []string{
-			FederatedServiceName,
-			fmt.Sprintf("%s.%s", FederatedServiceName, f.Namespace.Name),
-			fmt.Sprintf("%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name),
-		}
-		for _, name := range localSvcDNSNames {
-			discoverService(f, name, false)
-		}
+			svcDNSNames := []string{
+				FederatedServiceName,
+				fmt.Sprintf("%s.%s", FederatedServiceName, f.Namespace.Name),
+				fmt.Sprintf("%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name),
+				fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+				fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
+			}
+			// TODO(mml): This could be much faster.  We can launch all the test
+			// pods, perhaps in the BeforeEach, and then just poll until we get
+			// successes/failures from them all.
+			for _, name := range svcDNSNames {
+				discoverService(f, name)
+			}
+		})
 
-		svcDNSNames := []string{
-			fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
-			fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
-		}
-		for _, name := range svcDNSNames {
-			discoverService(f, name, true)
-		}
+		Context("non-local federated service", func() {
+			BeforeEach(func() {
+				framework.SkipUnlessFederated(f.Client)
+
+				// Delete a federated service shard in the default e2e Kubernetes cluster.
+				err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+			})
+
+			It("should be able to discover a non-local federated service", func() {
+				framework.SkipUnlessFederated(f.Client)
+
+				svcDNSNames := []string{
+					fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+					fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
```

Comment:
There are a few problems here, as discussed in person:

1. Deleting the service from the local cluster won't actually work, as the Federation Service Controller will re-create it (or at least it should, and if it doesn't, that's a bug).  Instead you'll have to remove all of the endpoints from the service (by deleting all pods that match the service's selector).  That should remove the DNS entry for the local shard of the federated service.  Again, if it doesn't, then that's a bug.
2. Even once the DNS records have been deleted,  DNS propagation will take at least TTL seconds (currently set to 180 in the Federated Service Controller) to propagate.  To get around that you'd probably need to resolve before deleting the service's backends, and then after deleting, until the IP address changes (which might take up to 180 seconds, due to DNS caching).  We might consider reducing that TTL in e2e test clusters to speed up the tests.